src/wallet/rpc/spend.cpp

```diff
@@ -88,8 +88,21 @@ std::set<int> InterpretSubtractFeeFromOutputInstructions(const UniValue& sffo_in
     return sffo_set;
 }
 
-static UniValue FinishTransaction(const std::shared_ptr<CWallet> pwallet, const UniValue& options, const CMutableTransaction& rawTx)
+static UniValue FinishTransaction(const std::shared_ptr<CWallet> pwallet, const UniValue& options, CMutableTransaction& rawTx)
 {
+    bool can_anti_fee_snipe = !options.exists("locktime");
+
+    for (const CTxIn& tx_in : rawTx.vin) {
+        // Can not be FINAL for locktime to work
+        can_anti_fee_snipe &= tx_in.nSequence != CTxIn::SEQUENCE_FINAL;
```

Comment:
In 8e5778a9fb6fbcbddd4ba87cc3f3caf41118592a (wallet, rpc: add anti-fee-sniping to `send` and `sendall`):

It seems to me that the `nSequence` check here is too restrictive. Locktime is disabled when _all_ sequences are set to final, but it seems to me that `can_anti_fee_snipe` would be set to `false` here, even if only one input’s sequence is set to final.

I’m not sure this would matter much in practice because most transactions would probably use the same sequence value for all inputs, but `can_anti_fee_snipe` should probably end up `true` if at least one `nSequence` is non-final?

Could you perhaps add a test that a transaction with final `nSequence` values does not get its `locktime` set, but one with mixed values and one with all non-final `nSequence` values both do?

Reply:
> It seems to me that the `nSequence` check here is too restrictive. Locktime is disabled when _all_ sequences are set to final, but it seems to me that `can_anti_fee_snipe` would be set to `false` here, even if only one input’s sequence is set to final.

Currently `DiscourageFeeSniping` will assert if any sequence is final, so this check is appropriate for now. It also enforces that the sequences are either `MAX_SEQUENCE_NONFINAL` or `MAX_BIP125_RBF_SEQUENCE`, so this needs to be checking that as well otherwise an assertion will be hit. Or a commit could be added to relax those checks.

> Could you perhaps add a test that a transaction with final nSequence values does not get its locktime set, but one with mixed values and one with all non-final nSequence values both do?

I would also like to see tests for other sequence numbers as there is a risk of reaching an assertion here.

Reply:
> It also enforces that the sequences are either MAX_SEQUENCE_NONFINAL or MAX_BIP125_RBF_SEQUENCE, so this needs to be checking that as well otherwise an assertion will be hit. Or a commit could be added to relax those checks.

It seems like this part of the comment wasn't addressed

Reply:
> It also enforces that the sequences are either MAX_SEQUENCE_NONFINAL or MAX_BIP125_RBF_SEQUENCE, so this needs to be checking that as well otherwise an assertion will be hit. Or a commit could be added to relax those checks

I have added these checks to `FinishTransaction`.